### PR TITLE
Add parameter -Group to Get-Verb

### PR DESCRIPTION
**Pull request by dchristian3188**
Adding group filter to get-verb. Please reference:
https://github.com/PowerShell/PowerShell/issues/2527

### CODE REVIEW

File: src/System.Management.Automation/engine/InitialSessionState.cs [hunk 1]
```diff
@@ -4751,7 +4751,11 @@ internal static string GetGetVerbText()
 param(
     [Parameter(ValueFromPipeline=$true)]
     [string[]]
-    $verb = '*'
+    $Verb = '*',
+
+    [Parameter(ValueFromPipeline=$true)]
+    [string[]]
+    $Group
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: I'd rather see enum here instead string.

Comment (dchristian3188) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: @iSazonov  What do you think would be the best way to accomplish this? Currently the groups are determined by looking at the class names and removing the word verb.

`[System.Reflection.IntrospectionExtensions]::GetTypeInfo([PSObject]).Assembly.ExportedTypes |
        Where-Object {$_.Name -match '^Verbs.'}`

I could put the current groups in a validate set but is that the best for long term support?

Comment (kittholland) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: DynamicParam maybe?

Comment (lzybkr) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: `ValidateSet` is reasonable, especially if there is an easy way to test.

I would like to see this command converted to C#.

Comment (iSazonov) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: You can see a enum using sample in [Get/Set-Alias](https://github.com/PowerShell/PowerShell/blob/309bf603f9aff9009dce7e725d42b2d4f99f58c9/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WriteAliasCommandBase.cs#L47)
The enum could be placed [ Verbs.cs](https://github.com/PowerShell/PowerShell/blob/309bf603f9aff9009dce7e725d42b2d4f99f58c9/src/System.Management.Automation/utils/Verbs.cs)

Comment (dchristian3188) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: thanks for the feedback. I'll get started on the changes

Comment (iSazonov) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: @lzybkr  Could you comment on the use of `enum` vs `validate set` ?
I attached the file with script and the resulting list of parameters whose type is enum (for Powershel 6 only).
[Test-Params.txt](https://github.com/PowerShell/PowerShell/files/618084/Test-Params.txt)

Comment (lzybkr) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: In this case, an enum introduces a new public api.
If the only use of that enum is for parameter validation and tab completion, then `ValidateSet` is might be better than introducing a new api.
If we did use an enum, the parameter doesn't need to be an array, the enum could use `Flags`.

File: src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1 [hunk 2]
```diff
@@ -2,7 +2,7 @@
 GUID="1DA87E53-152B-403E-98DC-74D7B4D63D59"
 Author="Microsoft Corporation"
 CompanyName="Microsoft Corporation"
-Copyright="� Microsoft Corporation. All rights reserved."
+Copyright="� Microsoft Corporation. All rights reserved."
```
Comment (lzybkr) [inline src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1]: Revert this change.

File: src/System.Management.Automation/engine/InitialSessionState.cs (hunk 1 above)
Comment (iSazonov) [inline src/System.Management.Automation/engine/InitialSessionState.cs]: @lzybkr Thanks for great comment! I was asking about using VerbGroups in Issue addressed  in the PR  but there were no examples. So I don't see the need for the public api at all.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            Get-Verb -Group FakeGroupNeverExists -ErrorAction Stop
+        }
+        Catch{
+            if($PSItem.FullyQualifiedErrorId -eq 'ParameterArgumentValidationError,Microsoft.PowerShell.Commands.GetVerbCommand'){
+                $result = $true
+            }
+            else{
+                $result = $false
+            }
+            $result | should be $true
```
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1]: More simply:

```PowerShell
$_.FullyQualifiedErrorId | Should Be ...
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    It "Should not return duplicate Verbs with Group paramater" -test {
+        $dupVerbs = Get-Verb -Group other,other
+        $uniqueVerbs = $dupVerbs | 
+            Select-Object -Unique
+        $dupVerbs.Count | Should be $uniqueVerbs.Count 
+    }
+
+    It "Should filter using the verb parameter" -test {
+        Get-Verb -Verb fakeVerbNeverExists | Should be ''
```
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1]: Shouldn't this be `... | Should BeNullOrEmpty`

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,48 @@
+﻿Describe "Get-Verb" -Tags "CI" {
+
+    It "Should get a list of Verbs" -test {
```
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1]: We don't use `-test` anywhere else, so it just looks weird here and should probably be removed.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    It "Should not return duplicate Verbs with Verb paramater" -test {
+        $dups = Get-Verb -Verb Add,ad*,a*
+        $unique = $dups | 
+            Select-Object -Unique
+        $dups.Count | Should be $unique.Count 
+    }
+
+    It "Should not return duplicate Verbs with Group paramater" -test {
+        $dupVerbs = Get-Verb -Group other,other
```
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1]: If the parameter was an enum with flags, you wouldn't need any special code to cover this case, we wouldn't see the group specified multiple times.

**Review by lzybkr — CHANGES_REQUESTED** [review]
Great tests!
I realize you mostly just rewrote the original function in C#, but I think the original implementation could have been better.

I realize the performance of this cmdlet really doesn't matter, but we have plenty of bad patterns that people just copy and I want to avoid introducing those patterns in addition to fixing the bad patterns over time.

**Comment by dchristian3188** [comment]
Got it, thanks for the review Izbkr. I'll get started on the changes.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,42 @@
+﻿Describe "Get-Verb" -Tags "CI" {
+
+    It "Should get a list of Verbs" {
+        Get-Verb | Should not be $null
+    }
+
+    It "Should get a specific verb" {
+        (Get-Verb -Verb Add | Measure-Object).Count | Should be 1
```
Comment (JamesWTruher) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1]: >(Get-Verb -Verb Add | Measure-Obj [](start = 8, length = 33)

this should be expressed as:
`@(Get-Verb -Verb Add).Count | should be 1`
no need to use measure-object (and the same for all of these uses of Measure-Object)

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1 [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        (Get-Verb -Verb Add -Group Common | Measure-Object).Count | Should be 1
+    }
+
+    It "Should get a specific group" {
+        Get-Verb -Group Common | Should not be $null
```
Comment (JamesWTruher) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1]: >-Verb -Group Common | Should not be $null [](start = 11, length = 41)

this doesn't actually validate that you got the proper group
i think you're looking for this:
`(Get-Verb -Group Common).Group | Sort-Object -uniq | Should be Common`

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        $dupVerbs.Count | Should be $uniqueVerbs.Count 
+    }
+
+    It "Should filter using the Verb parameter" {
+        Get-Verb -Verb fakeVerbNeverExists | Should BeNullOrEmpty
+    }
+
+    It "Should not accept Groups that are not in the validate set" {
+        try{
+            Get-Verb -Group FakeGroupNeverExists -ErrorAction Stop
```
Comment (JamesWTruher) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1]: >Get-Verb -Group FakeGroupNeverExists -ErrorAction Stop [](start = 12, length = 54)

in case this doesn't throw, you need to add:
`throw "Expected error did not occur"`
or similar

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1 (hunk 7 above)
Comment (dchristian3188) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1]: thanks! I was trying to use (Get-Verb -Verb Add).Count but was getting null results. Didn't think of adding the "@".

Comment (JamesWTruher) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Verb.Tests.ps1]: if you're not sure whether you'll get a singleton or a collection you can always add `@()`

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetVerbCommand.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+            foreach (Type type in verbTypes)
+            {
+                foreach (FieldInfo field in type.GetFields())
+                {
+                    if (field.IsLiteral)
+                    {
+                        if(this.Verb != null)
+                        {
+                            Collection<WildcardPattern> matchingVerbs = SessionStateUtilities.CreateWildcardsFromStrings(
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetVerbCommand.cs]: This creates a ton of garbage - move it outside the loops.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetVerbCommand.cs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                                WildcardOptions.IgnoreCase
+                            );
+                            
+                            if(!SessionStateUtilities.MatchesAnyWildcardPattern(field.Name,matchingVerbs,false))
+                            {
+                                continue;
+                            }
+                        }
+
+                        if(this.Group != null)
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetVerbCommand.cs]: This can be done once per type, not once per field in the type, correct?

**Review by lzybkr — CHANGES_REQUESTED** [review]
Looking pretty good now, just a couple more perf issues to fix.

As long as you're making these changes, you might as well use our consistent formatting, a space after `if` and spaces after arguments, instead of `foo(arg1,arg2)`, use `foo(arg1, arg2)`.

If you don't make those changes, we'll fix it later with a tool, but it's nicer to be in the habit of following the conventions of the project.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetVerbCommand.cs (hunk 11 above)
Comment (dchristian3188) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetVerbCommand.cs]: that makes alot of sense, I'll get this updated.

**Review by lzybkr — APPROVED** [review]
Looks good - please fix the merge conflict so we can merge the PR.

**Comment by kittholland** [comment]
Since you are using the validateset with hardcoded group names, can you add a test enumerating the possible groups and ensuring all possibilities are handled in the validateset?

File: src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1 [hunk 12]
```diff
@@ -2,7 +2,7 @@
 GUID="1DA87E53-152B-403E-98DC-74D7B4D63D59"
 Author="Microsoft Corporation"
 CompanyName="Microsoft Corporation"
-Copyright="� Microsoft Corporation. All rights reserved."
+Copyright="© Microsoft Corporation. All rights reserved."
```
Comment (lzybkr) [inline src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1]: This change should be reverted - it changes the file to UTF8 encoding without a BOM.

**Review by lzybkr — CHANGES_REQUESTED** [review]

**Comment by dchristian3188** [comment]
Added an additional test to cover extra groups. Please let me know if any other changes are needed.

File: src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1 (hunk 12 above)
Comment (lzybkr) [inline src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1]: Instead of reverting the change, you added a BOM.

That's a reasonable change to make (we have another PR to clean these up), but to keep this PR clean, I'd prefer you undo the change and not change the encoding.

**Review by lzybkr — APPROVED** [review]
